src/sideEffects/firebase.ts

```diff
@@ -26,80 +29,175 @@ function transformSnapshotIntoSensorData(
+export async function getSensor(
+  sensorId: string,
+  onSuccess: (param?: SensorData) => void,
   onError: (e: any) => void
 ) {
   try {
-    const sensorIds = Object.keys(place.sensors);
-    const promises = sensorIds.map((sensorId) =>
-      firebase.database().ref(`${refs.SENSORS}/${sensorId}`).once('value')
-    );
+    const sensorRef = db.collection('sensors').doc(sensorId);
+    const snapshot = await sensorRef.get();
+    const sensor = snapshot.data();
 
-    const sensors: { [sensorId: string]: SensorData } = {};
-    const results = await Promise.all(promises);
+    if (!sensor) {
+      onError({
+        code: 404,
+        status: 'Not Found',
+        message: 'Sensor not found',
+      });
+      return;
+    }
 
-    results.forEach((result) => {
-      const id = result.key;
-      const sensor: SensorSnapshot | null = result.val();
-      if (id && sensor) {
-        sensors[id] = transformSnapshotIntoSensorData(id, sensor);
-      }
+    // Adds datachain as as object field of chain objects
+    let datachain = {};
+    const datachainRef = sensorRef.collection('datachain');
+    const datachainSnapshot = await datachainRef.get();
+    datachainSnapshot.forEach((datachainItem) => {
+      const datachainDoc = datachainItem.data();
+      datachain = { ...datachain, [datachainItem.id]: datachainDoc };
     });
 
-    onSuccess(sensors);
+    // Adds systems as an object field of system objects
+    const systems = await getSystems();
+    const filteredSystems = {};
+    const { systemIds } = sensor;
+    systemIds.forEach((systemId: string | number) => {
+      filteredSystems[systemId] = systems[systemId];
+    });
+
+    const sensorWithDatachainAndSystems = {
+      ...sensor,
+      datachain,
+      systems: filteredSystems,
+    };
+    const sensorData = transformSnapshotIntoSensorData(
+      sensorId,
+      sensorWithDatachainAndSystems as SensorSnapshot
+    );
+    onSuccess(sensorData);
   } catch (e) {
     onError(e);
   }
 }
 
-export function getSensor(
-  sensorId: string,
-  onSuccess: (param?: SensorData) => void,
+// Gets all sensors in a place
+export async function getSensorsInPlace(
+  place: PlaceData,
+  onSuccess: (param: { [sensorId: string]: SensorData }) => void,
   onError: (e: any) => void
 ) {
   try {
-    const sensorRef = firebase.database().ref(`sensors/${sensorId}`);
-    sensorRef.on('value', async (snapshot) => {
-      if (!snapshot) {
-        onSuccess();
-        return;
-      }
+    const sensors: { [sensorId: string]: SensorData } = {};
+    const sensorsRef = db.collection('sensors');
+    const sensorsSnapshot = await sensorsRef
+      .where('placeId', '==', place.id)
+      .get();
 
-      const sensor: SensorSnapshot | null = snapshot.val();
-      if (!sensor) {
-        onError({
-          code: 404,
-          status: 'Not Found',
-          message: 'Sensor not found',
-        });
-        return;
-      }
+    sensorsSnapshot.forEach(async (sensorDoc) => {
```

Comment:
You can't await unside a forEach, even if you add async (not sure the point of the async here and why it's valid syntax, to be honest, so I can see why it would be confusing)

This is similar to the issue that was in the migration script, it awaits for nothing or nothing catches the await.

This works on second render because it doesn't need to hit the server so it returns it right away.

Two solutions here: https://stackoverflow.com/a/37576787

IMO there is a bigger issue here and likely why you're seeing it take so long, which is you are making database calls within a loop, ideally you would only query each collection once, which would mean either querying for only what you need, or over fetching and then mapping it together. It kind of seems like this loop is only being run once, so I can't say for sure this is an issue, but it is a "code smell" as they say. For `getSystems` you could make this call outside of the loop (assuming the loop is required) and then reference the results inside the loop if needed. Not sure about datachain.